pim6d: conversion work batch

eqvinox opened the pull request:
A lot of repetitive, mind-numbing adaptation work...

mobash-rasool commented on pimd/pim_msg.c:
```diff
@@ -97,6 +97,68 @@ uint8_t *pim_msg_addr_encode_ipv4_source(uint8_t *buf, struct in_addr addr,
 	return buf + PIM_ENCODED_IPV4_SOURCE_SIZE;
 }
 
+uint8_t *pim_msg_addr_encode_ipv6_source(uint8_t *buf, struct in6_addr addr,
```
Need to move the below functions under #if PIM_IPV == 6

eqvinox replied:
Nope, they work regardless of which protocol we're compiling for. We include IPv6 addrs in the secondary address list in IPv4 too…

qlyoung commented:
Can you fix the style issues?

mobash-rasool commented on pimd/pim_iface.c:
```diff
@@ -841,28 +832,42 @@ struct in_addr pim_find_primary_addr(struct interface *ifp)
-		if (p->family != AF_INET) {
+		switch (ifc->address->family) {
+		case AF_INET:
+			v4_addrs++;
```
This is now getting incremented for invalid ipv4 address (0.0.0.0) as well which differs from the previous behaviour. When IPv4 address is not configured, it tries to get the address from the loopback based on v4_addrs count outside of this loop. Please check.

rathnasabapathyv replied:
@mobash-rasool : I think, when IPv4 addr is not configured for an interface, then we will not have ifc->address->family set as AF_INET for that case. I think v4_addrs will not increment and it will remain 0 for IP fetch from loopback intf down below.

eqvinox replied:
The `0.0.0.0` address check there is bogus / cargo cult to begin with. It's probably carried over from some old code that used `ioctl` based kernel APIs where you get 0.0.0.0 if there is no IPv4 address configured. That is not the case with zebra, we get a list of addresses, and if there is no address the list is empty. There will never be a 0.0.0.0 address.

I've removed the check since it's bogus. The code overall is correct, even if it doesnt 100% match the previous code.

(btw: this is why this PR is separate from #10292 - in #10292 I tried to only do 100% identical changes.)

mobash-rasool reviewed (changes requested):
I have few comments. Please check.

rathnasabapathyv commented on pimd/pim_iface.c:
```diff
@@ -401,19 +394,18 @@ static int pim_sec_addr_update(struct interface *ifp)
-		}
 
-		if (pim_ifp->primary_address.s_addr == p->u.prefix4.s_addr) {
+		if (pim_addr_cmp(addr, pim_ifp->primary_address)) {
```
"!" check missing here for prim-addr equal comparision & skip

rathnasabapathyv replied:
All other changes LGTM.

eqvinox replied:
Good catch! Fixed.

eqvinox commented on pimd/pim_addr.h:
```diff
@@ -21,19 +21,88 @@
+	pim_addr ret;
+
+	if (in.p->family != PIM_AF)
+		return PIMADDR_ANY;
```
This code was changed in #10292 and isn't touched in this PR, I'll leave this as it is to not intermix things further. It can always be changed in a separate PR.

(I'm also wondering if it should be an `assert()`; but the question there is whether it is preferable to crash the daemon or continue with a zero address...)

eqvinox commented on pimd/pim_tlv.c:
```diff
@@ -532,40 +557,33 @@ int pim_parse_addr_group(pim_sgaddr *sg, const uint8_t *buf, int buf_size)
+
+	if ((addr + sizeof(sg->grp)) > pastend) {
+		zlog_warn(
+			"%s: IPv4 group address overflow: left=%td needed=%zu from",
```
fixed.

eqvinox commented:
- rebased to current `frr/master`
- addressed review comments
- fixed style

The 1 remaining style issue is clang version dependent. I used `Debian clang-format version 13.0.1-+rc3-1~exp1` and this is what it gave me. We can't go chase minute differences in space characters that are caused by the clang version being *sliiiightly* different, we'll never be done :disappointed:

Also sorry for the delay on this, I had/have some health issues going on.

mobash-rasool commented:
ci:rerun

davischw commented:
ci:rerun

davischw commented:
ci:rerun
